splitgraph/ingestion/sqlite/__init__.py

```diff
@@ -0,0 +1,177 @@
+    yield local_fh.name
+    os.remove(local_fh.name)
+
+
+def query_connection(
+    con: sqlite3.Connection, sql: str, parameters: Optional[Dict[str, str]] = None
+) -> List[Any]:
+    with contextlib.closing(con.cursor()) as cursor:
+        cursor.execute(sql, parameters or {})
+        return cursor.fetchall()
+
+
+@contextmanager
+def db_from_minio(url: str) -> Generator[sqlite3.Connection, None, None]:
+    with minio_file(url) as f:
+        with contextlib.closing(sqlite3.connect(f)) as con:
+            yield con
+
+
+def sql_quote_str(s: str) -> str:
+    return s.replace("'", "''").replace('"', '""')
+
+
+# partly based on https://stackoverflow.com/questions/1942586/comparison-of-database-column-types-in-mysql-postgresql-and-sqlite-cross-map
+def sqlite_to_postgres_type(sqlite_type: str) -> str:
+    if sqlite_type == "DATETIME":
+        return "TIMESTAMP WITHOUT TIME ZONE"
+    # from: https://www.sqlite.org/datatype3.html#determination_of_column_affinity
+    # If the declared type contains the string "INT" then it is assigned INTEGER affinity.
+    if "INT" in sqlite_type:
+        return "INTEGER"
+    # If the declared type of the column contains any of the strings "CHAR", "CLOB", or "TEXT" then that column has TEXT affinity. Notice that the type VARCHAR contains the string "CHAR" and is thus assigned TEXT affinity.
+    if "CHAR" in sqlite_type or "CLOB" in sqlite_type or "TEXT" in sqlite_type:
+        return "TEXT"
+    # If the declared type for a column contains the string "BLOB" or if no type is specified then the column has affinity BLOB.
+    if "BLOB" in sqlite_type:
+        return "BLOB"
+    # If the declared type for a column contains any of the strings "REAL", "FLOA", or "DOUB" then the column has REAL affinity.
+    if "REAL" in sqlite_type or "FLOA" in sqlite_type or "DOUB" in sqlite_type:
+        return "REAL"
+    # Otherwise, the affinity is NUMERIC. TODO: Precision and scale
+    return "NUMERIC"
+
+
+def sqlite_connection_to_introspection_result(con: sqlite3.Connection) -> IntrospectionResult:
+    schema: IntrospectionResult = cast(IntrospectionResult, {})
+    for (
+        table_name,
+        column_id,
+        column_name,
+        column_type,
+        _notnull,
+        _default_value,
+        pk,
+    ) in query_connection(con, LIST_TABLES_QUERY, {}):
+        table = schema.get(table_name, ([], TableParams({})))
+        cast(Tuple[List[TableColumn], TableParams], table)[0].append(
+            TableColumn(column_id + 1, column_name, sqlite_to_postgres_type(column_type), pk != 0)
+        )
+        schema[table_name] = table
+    return schema
+
+
+class SQLiteDataSource(LoadableDataSource):
+
+    table_params_schema: Dict[str, Any] = {"type": "object", "properties": {}}
+
+    params_schema: Dict[str, Any] = {
+        "type": "object",
+        "properties": {
+            "url": {
+                "type": "string",
+                "description": "HTTP URL to the sqlite file",
```

Comment:
(given that this is going to be rendered in the UI)

```suggestion
                "description": "HTTP URL to the SQLite file",
```